File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    // input is an array of nums
    // output is max profit or 0
    // prices =[7,1,5,3,6,4]
    // no negatives, length can be 1

    // while iterating determine when to buy
    // and determine when to sell


    let maxPrice = 0
    let l = 0
    let r = 1
    while (r < prices.length) {
        if (prices[r] < prices[l]) {
            l = r
            r++
        } else {
            profit = prices[r] - prices[l]
            maxPrice = Math.max(maxPrice, profit)
            r++
        }
    }
    return maxPrice

};

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
    # Input: prices[] nums
    # Output: will be either 0 or profit
    # If prices is sorted and is in descending order, then we return 0
    profit = 0
    l = 0
    r = 0
    while l <= r && r < prices.length
        if prices[r] <= prices[l]
            l = r
            r += 1
            next
        else
            profit = [profit, prices[r] - prices[l]].max
        end
        r += 1
    end
    profit
end